scheduler/cmd/scheduler/main.go

```diff
@@ -17,21 +17,90 @@
 package main
 
 import (
+	"flag"
+	"fmt"
 	"os"
+	"strings"
 
 	"k8s.io/component-base/cli"
-	_ "k8s.io/component-base/metrics/prometheus/clientgo" // for rest client metric registration
-	_ "k8s.io/component-base/metrics/prometheus/version"  // for version metric registration
+	_ "k8s.io/component-base/logs/json/register" // for JSON log format registration
+	_ "k8s.io/component-base/metrics/prometheus/clientgo"
+	_ "k8s.io/component-base/metrics/prometheus/version" // for version metric registration
 	"k8s.io/kubernetes/cmd/kube-scheduler/app"
+	"k8s.io/kubernetes/pkg/scheduler/apis/config"
+	"k8s.io/kubernetes/pkg/scheduler/apis/config/scheme"
 
 	wasm "sigs.k8s.io/kube-scheduler-wasm-extension/scheduler/plugin"
 )
 
 func main() {
-	command := app.NewSchedulerCommand(
-		app.WithPlugin(wasm.PluginName, wasm.New),
-	)
+	pluginNames, err := getWasmPluginsFromConfig()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to get wasm plugins from config: %v\n", err)
+		os.Exit(1)
+	}
+
+	opt := []app.Option{}
+	for _, pluginName := range pluginNames {
+		opt = append(opt, app.WithPlugin(pluginName, wasm.PluginFactory(pluginName)))
+	}
+
+	command := app.NewSchedulerCommand(opt...)
 
 	code := cli.Run(command)
 	os.Exit(code)
 }
+
+// getWasmPluginsFromConfig parses the scheduler configuration specified with --config option,
+// and return the wasm plugins enabled by the user.
+func getWasmPluginsFromConfig() ([]string, error) {
+	// In the scheduler, the path to the scheduler configuration is specified with --config option.
+	configFile := flag.String("config", "", "")
+	flag.Parse()
+
+	if configFile == nil {
+		// Users don't have the own configuration. do nothing.
+		return nil, nil
+	}
+
+	cfg, err := loadConfigFromFile(*configFile)
+	if err != nil {
+		return nil, err
+	}
+
+	return getWasmPluginNames(cfg), nil
+}
+
+// getWasmPluginNames returns the wasm plugin names enabled by the user.
+// It assumes that the wasm plugin is specified as the multi-point plugin,
+// and the wasm plugin name starts with wasm.PluginNamePrefix.
+func getWasmPluginNames(cc *config.KubeSchedulerConfiguration) []string {
+	wasmplugins := []string{}
+	for _, profile := range cc.Profiles {
+		// assuming that the wasm plugin is specified as a multi-point plugin.
+		for _, plugin := range profile.Plugins.MultiPoint.Enabled {
```

Comment:
Is this enabled by default? It is cool that there is a concept of having multiple plugins, but I wonder why we would do this which requires a name prefix?

If we don't use "multipoint" we could use another nested property, allow the name to be plain, and not accidentally turned off. For example, look for guestPath instead and don't try to fit it into the MultiPoint thing.

Then, everything in the plugin is handled normally, it seems. What am I missing?

Reply:
Removed a name prefix.